### PR TITLE
Add IndexConfiguration CRUD to IndexManager

### PR DESCRIPTION
**Pull request by wu-hui**
#no-changelog

### CODE REVIEW

File: Firestore/core/test/unit/local/leveldb_index_manager_test.cc [hunk 1]
```diff
@@ -38,6 +44,75 @@ INSTANTIATE_TEST_SUITE_P(LevelDbIndexManagerTest,
...
+      EXPECT_EQ(indexes.size(), 1);
+    }
+
+    index_manager->DeleteFieldIndex(index);
+    {
+      auto indexes = index_manager->GetFieldIndexes("coll1");
+      EXPECT_EQ(indexes.size(), 0);
+    }
+  });
+}
```
Comment (schmidt-sebastian) [inline Firestore/core/test/unit/local/leveldb_index_manager_test.cc:114]: Please add a test that verifies that indexes are loaded during startup (similar to Android's `testChangeUser()`).
You might also want to implement updateCollectionGroup and nextCollectionGroup to update so you can verify the priority queue - but I am okay with doing this in a follow up.

File: Firestore/core/src/local/leveldb_index_manager.cc [hunk 2]
```diff
@@ -71,6 +125,218 @@ std::vector<ResourcePath> LevelDbIndexManager::GetCollectionParents(
...
+    }
+  }
+
+  started_ = true;
+}
+
+void LevelDbIndexManager::DeleteFromUpdateQueue(FieldIndex* index_ptr) {
+  // Pop and save `FieldIndex*` until index_ptr is found, then pushed what are
+  // popped out back to `next_index_to_update_` except for `index_ptr`.
+  std::vector<FieldIndex*> popped_out;
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply Firestore/core/src/local/leveldb_index_manager.cc:197]: You are correct - I thought Java used a tree here which would have made the removal `O(log n)`. That is not the case though.

File: Firestore/core/src/local/index_manager.h [hunk 3]
```diff
@@ -58,6 +71,65 @@ class IndexManager {
    */
   virtual std::vector<model::ResourcePath> GetCollectionParents(
       const std::string& collection_id) = 0;
+
+  /**
+   * Adds a field path index.
+   *
+   * The actual entries for this index will be created and persisted
+   * in the background by the SDK, and the index will be used for query
+   * execution once values are persisted.
```
Comment (schmidt-sebastian) [inline Firestore/core/src/local/index_manager.h:80]: NIt: This looks like it is not using the full line length.

File: Firestore/core/src/local/leveldb_index_manager.cc [hunk 4]
```diff
@@ -71,6 +125,229 @@ std::vector<ResourcePath> LevelDbIndexManager::GetCollectionParents(
...
+
+      auto segments = serializer_->DecodeFieldIndexSegments(&reader, *message);
+      if (!reader.ok()) {
+        HARD_FAIL("Index proto failed to decode: %s",
+                  reader.status().ToString());
+      }
+
+      // If we fetched an index state for the user above, combine it with this
+      // index. We use the default state if we don't have an index state (e.g.
+      // the index was created while a different user as logged in).
```
Comment (schmidt-sebastian) [inline Firestore/core/src/local/leveldb_index_manager.cc:177]:
```suggestion
      // the index was created while a different user was logged in).
```

File: Firestore/core/src/local/leveldb_index_manager.cc [hunk 5]
```diff
@@ -71,6 +125,229 @@ std::vector<ResourcePath> LevelDbIndexManager::GetCollectionParents(
...
+      auto iter = index_states.find(config_key.index_id());
+      IndexState state = iter != index_states.end()
+                             ? iter->second
+                             : FieldIndex::InitialState();
+
+      // Store the index and update `memoized_max_index_id` and
+      // `memoized_max_sequence_number`.
```
Comment (schmidt-sebastian) [inline Firestore/core/src/local/leveldb_index_manager.cc:184]:
```suggestion
      // Store the index and update `memoized_max_index_id_` and
      // `memoized_max_sequence_number_`.
```

File: Firestore/core/src/local/leveldb_index_manager.cc [hunk 6]
```diff
@@ -71,6 +125,229 @@ std::vector<ResourcePath> LevelDbIndexManager::GetCollectionParents(
...
+
+  auto existing_index_iter = existing_indexes.find(index_id);
+
+  if (existing_index_iter != existing_indexes.end()) {
+    DeleteFromUpdateQueue(&existing_index_iter->second);
+  }
+
+  // Moves `index` into `existing_indexes`.
+  existing_indexes.insert({index_id, std::move(index)});
+  // next_index_to_update_ holds a pointer to Index owned by `existing_indexes`.
```
Comment (schmidt-sebastian) [inline Firestore/core/src/local/leveldb_index_manager.cc:229]:
```suggestion
  // next_index_to_update_ holds a pointer to the index owned by `existing_indexes`.
```

Maybe

File: Firestore/core/src/local/leveldb_index_manager.h [hunk 7]
```diff
@@ -52,6 +94,27 @@ class LevelDbIndexManager : public IndexManager {
...
+  QueueForNextIndexToUpdate next_index_to_update_;
+  int32_t memoized_max_index_id_ = -1;
+  int64_t memoized_max_sequence_number_ = -1;
+
+  /* Owned by LevelDbPersistence. */
+  LocalSerializer* serializer_ = nullptr;
+
+  bool started_ = false;
+
+  std::string uid_;
```
Comment (schmidt-sebastian) [inline Firestore/core/src/local/leveldb_index_manager.h:117]: This is not initialized, is it?

File: Firestore/core/src/local/leveldb_key.cc [hunk 8]
```diff
@@ -1151,10 +1167,12 @@ std::string LevelDbIndexConfigurationKey::KeyPrefix() {
   return writer.result();
 }
 
-std::string LevelDbIndexConfigurationKey::Key(int32_t id) {
+std::string LevelDbIndexConfigurationKey::Key(
+    int32_t id, absl::string_view collection_group) {
   Writer writer;
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply Firestore/core/src/local/leveldb_key.cc:1172]: I will play around with it if I have some time, but let's keep it for now.

**Review by schmidt-sebastian — APPROVED** [review]

File: Firestore/core/src/local/index_manager.h (hunk 3 above)
Comment (wu-hui) [reply Firestore/core/src/local/index_manager.h]: Fixed.

File: Firestore/core/src/local/leveldb_index_manager.h (hunk 7 above)
Comment (wu-hui) [reply Firestore/core/src/local/leveldb_index_manager.h:117]: It is? https://en.cppreference.com/w/cpp/string/basic_string/basic_string

Default constructor creates an empty string.

Comment (schmidt-sebastian) [reply Firestore/core/src/local/leveldb_index_manager.h:117]: Ack.

File: Firestore/core/src/local/local_serializer.cc [hunk 9]
```diff
@@ -453,6 +458,87 @@ BundledQuery LocalSerializer::DecodeBundledQuery(
...
+LocalSerializer::EncodeFieldIndexSegments(
+    const std::vector<model::Segment>& segments) const {
+  Message<google_firestore_admin_v1_Index> result;
+
+  result->query_scope =
+      google_firestore_admin_v1_Index_QueryScope_COLLECTION_GROUP;
+
+  result->fields_count = segments.size();
+  result->fields =
+      MakeArray<google_firestore_admin_v1_Index_IndexField>(segments.size());
```
Comment (ncooke3) [inline Firestore/core/src/local/local_serializer.cc:503]: Looks like these new lines caused a few warnings. Reported in https://github.com/firebase/firebase-ios-sdk/issues/9430

Comment (dconeybe) [reply Firestore/core/src/local/local_serializer.cc:503]: Thank you for reporting this. #9433 fixes the warnings.

**Pull request merged**